demo/src/reader.rs

```diff
@@ -42,29 +42,29 @@ impl ReadError {
     }
 }
 
-trait SeekableRead: io::Read + io::Seek {}
-impl<T: io::Read + io::Seek> SeekableRead for T {}
+trait SeekableRead<'a>: io::Read + io::Seek + 'a {}
+impl<'a, T: io::Read + io::Seek + 'a> SeekableRead<'a> for T {}
 
-pub struct Reader {
-    data: Box<dyn SeekableRead>,
+pub struct Reader<'a> {
+    data: Box<dyn SeekableRead<'a>>,
```

Comment:
```suggestion
    data: Box<dyn SeekableRead + 'a>,
```
I *believe* changing the trait itself isn't necessary, the `+ 'a` here should suffice.

Reply:
Interesting. Could you elaborate how this solved the weird lifetime issue in `tools/demo_read_write`?
(I unfortunately already squashed the commits. From memory: The writer was initialized with references from the reader. Then the borrow checker complained when the writer was used later)

Reply:
There are a couple of ways you can recover squashed commits:
- In your `git reflog`.
- In this pull request, ctrl-f "force-pushed".
- Using GitHub events if it's only on a remote branch: https://neodyme.io/en/blog/github_secrets/.

I couldn't quickly see why the code wasn't accepted before. `SeekableRead<'a>` is different from `SeekableRead + 'a` because the compiler cannot assume that a type also implements `SeekableRead<'b>` if it implements `SeekableRead<'a>` and `'b` is shorter than `'a`. But of course every type that is `SeekableRead + 'a` is also `SeekableRead + 'b`.

Reply:
The only change here is that the lifetime is also in the trait.
The lifetime of the trait is the same as the `file` parameter. The other parameters, which cause the issue here, shouldn't be taken into account afaik